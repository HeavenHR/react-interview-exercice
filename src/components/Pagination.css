:local(.pageStyle) {
  padding: 5px;
  text-align: center;
}

:local(.pageStyle button) {
  background: none;
  font-size: 1em;
  margin: 2px;
  color: white;
  border: 1px solid white;
}

:local(.pageStyle button:hover) {
  color: #5c75b0;
  background: white;
  border: 1px solid white;
}

:local(.pageStyle span) {
  color: white;
  margin: 0 4px 0 4px;
} 

:local(.pageStyle a) {
  display: inline-block;
  padding: 6px 12px;
  margin-bottom: 0;
  font-size: 14px;
  font-weight: 400;
  line-height: 1.42857143;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  -ms-touch-action: manipulation;
  touch-action: manipulation;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 1px solid transparent;
  border-radius: 4px;
  border: 1px solid white;
  margin: 2px;
  color: white;
}

:local(.pageStyle a:hover) {
  color: #5c75b0;
  background: white;
  border: 1px solid white;
  text-decoration: none;
}

:local(.pageStyle a:active) {
  color: white;
  background: #5c75b0;
  border: 1px solid white;
  text-decoration: none;
}
