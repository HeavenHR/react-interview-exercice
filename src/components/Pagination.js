import React, { Component, PropTypes } from 'react';

class Pagination extends Component {

  render () {
    const { pagination, size, onUpdatePage } = this.props;

    if (pagination.itemsPerPage >= size) {
      return null;
    }

    const pages = this.createPages();

    return (
      <div>
        {pages.map(page => (
          <button key={`page:${page}`} 
                  onClick={() => onUpdatePage(page)} 
                  disabled={pagination.page === page}>
                  {page}
          </button>
        ))}
      </div>
    );
  }

  createPages () {
    const numberOfPages = Math.round(this.props.size / this.props.pagination.itemsPerPage);
    const pages = [...new Array(numberOfPages)];

    return pages.map((_, index) => index + 1);
  }
}

Pagination.propTypes = {
  size: PropTypes.number.isRequired,
  pagination: PropTypes.shape({
    page: PropTypes.number,
    itemsPerPage: PropTypes.number
  }),
  onUpdatePage: PropTypes.func.isRequired
};

export default Pagination;
