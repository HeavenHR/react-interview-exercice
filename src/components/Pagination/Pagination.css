:local(.paginationWrapper) {
  display: flex;
  justify-content: center;
}

:local(li.current a) {
  background-color: #F4F3F0;
}

:local(li.page-item a):active {
  background-color: #fff;
  color: #337ab7;
}