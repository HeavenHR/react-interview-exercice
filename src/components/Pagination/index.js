import React, { Component } from 'react';

import styles from './styles.css';

const time = new Date().getTime();

class Pagination extends Component {
  constructor(props) {
    super(props);
    
    this.state = {
      pageIndexChanged: false
    };

    this.handleChange = this.handleChange.bind(this);
    this.handlePrevButtonClick = this.handlePrevButtonClick.bind(this);
    this.handleNextButtonClick = this.handleNextButtonClick.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.selectedPageIndex !== this.props.selectedPageIndex) {
      this.setState(prevProps => ({
        pageIndexChanged: !prevProps.pageIndexChanged
      }));
    }
  }

  handleChange(e) {
    const pageIndex = Number(e.currentTarget.value);
    
    this.props.onPageSelect(pageIndex);
  }

  handlePrevButtonClick(e) {
    this.props.onPageSelect(this.props.selectedPageIndex - 1);
  }

  handleNextButtonClick(e) {
    this.props.onPageSelect(this.props.selectedPageIndex + 1);
  }

  render() {
    const { friendsListByChunks = [], selectedPageIndex = 0 } = this.props;
    const pagesCount = friendsListByChunks.length;

    return(
      <div className={`${styles.container}`}>
        {pagesCount > 1 && (
          <div className={`${styles.buttonGroup}`}>
            {/* disabled-attribute must be provided to disable the button on edge cases */}
            <button type="button" className="btn" onClick={this.handlePrevButtonClick} disabled={selectedPageIndex === 0} ><span className="glyphicon glyphicon-chevron-left" /></button>
            <select className={`${styles.select}`} value={selectedPageIndex} onChange={this.handleChange}>
              {friendsListByChunks.map((friends, key) => (
                /**
                 * Using "time = new Date().getTime()" to create unique-keys, usually any item from the API-response will be having unique_id,
                 *  then there is no need to construct unique-keys like this.
                 */
                <option key={`${time}_${key}`} value={key}>page {key + 1} of {pagesCount}</option>
              ))}
            </select>
            {/* disabled-attribute must be provided to disable the button on edge cases */}
            <button type="button" className="btn" onClick={this.handleNextButtonClick} disabled={selectedPageIndex === pagesCount - 1}><span className="glyphicon glyphicon-chevron-right" /></button>
          </div>
        )}
      </div>
    );
  }
}

export default Pagination;
