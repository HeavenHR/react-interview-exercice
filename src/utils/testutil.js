import { mount, shallow } from 'enzyme';
import { prettyPrint } from 'html';
import * as types from '../constants/ActionTypes';

export const mockFunction = jest.fn(value => value);

export const sampleFriend = {
  name: 'Sample friend',
  gender: 'male'
};

export const sampleId = 1;

export const createMountedComponent = ({ children, isShallow }) => {
  const mountedComponent = isShallow ? shallow(children) : mount(children);
  mountedComponent.htmlRender = () => prettyPrint(mountedComponent.html(), { indent_size: 2 });

  return mountedComponent;
};

export const inputProps = {
  value: 'sample name',
  onSubmit: mockFunction,
  onChange: mockFunction,
  isInvalid: false,
  errorMessage: 'Please fill friend name'
};

export const buttonGroupProps = {
  selected: '',
  options: [
    { value: 'male', showIcon: true },
    { value: 'female', showIcon: true }
  ],
  onClick: mockFunction,
  isInvalid: false,
  errorMessage: 'Please select gender'
};

export const changeEventObject = { target: { value: 'sample text' } };

export const samplePages = [1, 2, 3, 4, 5, 6, 7];

export const totalPages = samplePages.length;

export const addFriendAction = {
  type: types.ADD_FRIEND,
  friend: sampleFriend
};

export const expectedAddFriendState = {
  friendsById: [
    {
      name: 'Sample friend',
      gender: 'male'
    },
    {
      name: 'Theodore Roosevelt',
      starred: true,
      gender: 'male'
    },
    {
      name: 'Abraham Lincoln',
      starred: false,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const deleteFriendAction = {
  type: types.DELETE_FRIEND,
  id: 0
};

export const expectedDeleteFriendState = {
  friendsById: [
    {
      name: 'Abraham Lincoln',
      starred: false,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const starFriendAction = {
  type: types.STAR_FRIEND,
  id: 1
};

export const expectedStarFriendState = {
  friendsById: [
    {
      name: 'Theodore Roosevelt',
      starred: true,
      gender: 'male'
    },
    {
      name: 'Abraham Lincoln',
      starred: true,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const updatePaginationAction = {
  type: types.UPDATE_PAGINATION,
  currentPage: 2
};

export const expectedUpdatedPaginationState = {
  pageSize: 2,
  currentPage: 2,
  pagingRange: 3
};
