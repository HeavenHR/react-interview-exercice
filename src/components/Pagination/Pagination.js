import React, { Component } from 'react';
import PropTypes from 'prop-types';
import range from 'lodash/range';
import classNames from 'classnames/bind';
import styles from './Pagination.css';
import { derivePageItems, deriveNextPage } from '../../utils/pagination';

const cn = classNames.bind(styles);

class Pagination extends Component {
  static propTypes = {
    totalPages: PropTypes.number.isRequired,
    currentPage: PropTypes.number.isRequired,
    pagingRange: PropTypes.number.isRequired,
    updatePagination: PropTypes.func
  }

  static defaultProps = {
    totalPages: 6,
    currentPage: 1,
    pagingRange: 3
  }

  state = {
    totalPages: this.props.totalPages,
    currentPage: this.props.currentPage,
    pagingRange: this.props.pagingRange
  }

  componentWillReceiveProps(nextProps) {
    const { totalPages, pagingRange } = nextProps;

    if (this.props !== nextProps) {
      this.setState({ totalPages, pagingRange });
    }
  }

  handlePageSelection = newPage => () => {
    const { totalPages, currentPage } = this.state;
    const { updatePagination } = this.props;
    const nextPage = deriveNextPage(newPage, currentPage, totalPages);

    if (nextPage !== currentPage) {
      this.setState({ currentPage: nextPage });
      updatePagination && updatePagination(nextPage);
    }
  }

  renderPagination = () => {
    const { totalPages, currentPage, pagingRange } = this.state;
    const pageNumbers = range(1, totalPages + 1);
    const pageList = derivePageItems(pageNumbers, Math.ceil(currentPage / pagingRange), pagingRange);

    const listItem = ['< prev', ...pageList, 'next >'];
    const isDisabled = page => (currentPage === 1 && page === '< prev') ||
      (currentPage === totalPages && page === 'next >');

    return listItem.map((page, index) =>
      <li
        onClick={this.handlePageSelection(page)}
        key={index}
        className={cn('page-item', { current: page === currentPage, disabled: isDisabled(page) })}
      >
        <a className={cn('page-link')} href="#">{page}</a>
      </li>
    );
  }

  render() {
    return (
      <div className={cn(styles.paginationWrapper)}>
        <ul className={cn('pagination')}>
          {this.renderPagination()}
        </ul>
      </div>
    );
  }
}

export default Pagination;
