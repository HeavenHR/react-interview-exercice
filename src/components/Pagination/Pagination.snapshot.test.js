import React from 'react';

import Pagination from './Pagination';
import { createMountedComponent, mockFunction } from '../../utils/testutil';


describe('<Pagination /> should', () => {
  let mountedPagination;
  let pages;

  beforeEach(() => {
    mountedPagination = createMountedComponent({
      children: <Pagination updatePagination={mockFunction} />
    });

    pages = mountedPagination.find('li.page-item').map(item => item);
  });

  it('render correctly according to props', () => {
    expect(mountedPagination.htmlRender()).toMatchSnapshot();
    expect(pages.length).toEqual(5);
    expect(pages[0].hasClass('disabled')).toBeTruthy();
  });

  it('render next set of pages, when currentPage > pageRange', () => {
    pages[4].simulate('click');
    pages[4].simulate('click');
    pages[4].simulate('click');
    pages[4].simulate('click');

    expect(mountedPagination.htmlRender()).toMatchSnapshot();
    expect(mountedPagination.props().updatePagination).toBeCalledWith(2);
    expect(mountedPagination.props().updatePagination).toBeCalledWith(3);
    expect(mountedPagination.props().updatePagination).toBeCalledWith(4);
    expect(mountedPagination.props().updatePagination).toBeCalledWith(5);
    expect(mountedPagination.state().currentPage).toEqual(5);
  });
});