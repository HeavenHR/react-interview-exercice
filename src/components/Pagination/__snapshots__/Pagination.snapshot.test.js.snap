exports[`<Pagination /> should render correctly according to props 1`] = `
"<div class=\"\">
  <ul class=\"pagination\">
    <li class=\"page-item disabled\"><a class=\"page-link\" href=\"#\">&lt; prev</a>
    </li>
    <li class=\"page-item current\"><a class=\"page-link\" href=\"#\">1</a>
    </li>
    <li class=\"page-item\"><a class=\"page-link\" href=\"#\">2</a>
    </li>
    <li class=\"page-item\"><a class=\"page-link\" href=\"#\">3</a>
    </li>
    <li class=\"page-item\"><a class=\"page-link\" href=\"#\">next &gt;</a>
    </li>
  </ul>
</div>"
`;

exports[`<Pagination /> should render next set of pages, when currentPage > pageRange 1`] = `
"<div class=\"\">
  <ul class=\"pagination\">
    <li class=\"page-item\"><a class=\"page-link\" href=\"#\">&lt; prev</a>
    </li>
    <li class=\"page-item\"><a class=\"page-link\" href=\"#\">4</a>
    </li>
    <li class=\"page-item current\"><a class=\"page-link\" href=\"#\">5</a>
    </li>
    <li class=\"page-item\"><a class=\"page-link\" href=\"#\">6</a>
    </li>
    <li class=\"page-item\"><a class=\"page-link\" href=\"#\">next &gt;</a>
    </li>
  </ul>
</div>"
`;
